### PR TITLE
ICMSLST-1436 - PDF pre-sign/preview view protection

### PR DESCRIPTION
**Pull request by chris-pettinga**
ICMSLST-1436

Checking application status to protect against pre-sign PDFs being generated when they're not warranted

### CODE REVIEW

File: web/tests/conftest.py [hunk 1]
```diff
@@ -576,6 +575,25 @@ def fa_sil_app_in_progress(
...
+@pytest.fixture()
+def fa_sil_app_processing(
+    importer_client, importer, office, importer_one_contact, ilb_admin_user
+) -> SILApplication:
+    app = create_in_progress_fa_sil_app(importer_client, importer, office, importer_one_contact)
+    submit_app(client=importer_client, view_name=app.get_submit_view_name(), app_pk=app.pk)
+    app.status = ImpExpStatus.PROCESSING
+    app.case_owner = ilb_admin_user
+    app.save()
+
```
Comment (mikeylockwood) [inline web/tests/conftest.py:587]: Call `ilb_admin_client.post(CaseURLS.take_ownership(app.pk))` to set an app to `PROCESSING`. See `completed_sil_app` fixture

File: web/tests/conftest.py [hunk 2]
```diff
@@ -576,6 +575,25 @@ def fa_sil_app_in_progress(
...
+    previous_task = case_progress.get_expected_task(app, Task.TaskType.PROCESS)
+    Task.objects.create(process=app, task_type=Task.TaskType.AUTHORISE, previous=previous_task)
+
+    case_progress.check_expected_status(app, [ImpExpStatus.PROCESSING])
+    case_progress.check_expected_task(app, Task.TaskType.AUTHORISE)
```
Comment (mikeylockwood) [inline web/tests/conftest.py:592]: Call `ilb_admin_client.post(CaseURLS.start_authorisation(app.pk))`. See `completed_sil_app` fixture

**Review by mikeylockwood — COMMENTED** [review]
Apps in `test_fa_sil_pre_sign_licence_view`, `test_fa_oil_pre_sign_licence_view`, `test_fa_dfl_pre_sign_licence_view` need to be put into the correct state for pre sign with

`ilb_admin_client.post(CaseURLS.take_ownership(app.pk))` `ilb_admin_client.post(CaseURLS.start_authorisation(app.pk))`

File: web/domains/case/views/views_pdf.py [hunk 3]
```diff
@@ -42,6 +44,26 @@ def get(self, request: AuthenticatedHttpRequest, *args, **kwargs) -> HttpRespons
...
+                application=self.application,
+                tasks=case_progress.get_active_task_list(self.application),
+                app_checker=AppChecker(request.user, self.application),
+            )
+            if not action_config.show_link():
+                # The user is requesting a pre_sign version of the document, but they are not able to authorise it.
+                # reject them.
+                raise PermissionDenied()
+
         if document_type not in self.document_types:
```
Comment (MattHolmes123) [inline web/domains/case/views/views_pdf.py:67]: I don't really like this, it achieves the correct output but I don't like using the workbasket actions in this way.

A more explicit solution could be to create specific views, It wouldn't actually take a big change:
```
class PreviewLicenceView(DocumentPreviewBase):
    document_types = [DocumentTypes.LICENCE_PREVIEW, DocumentTypes.LICENCE_PRE_SIGN]
    output_filename = "Licence-Preview.pdf"


# This is new
class PresignLicenceView(PreviewLicenceView):
    current_status = [ImpExpStatus.PROCESSING, ImpExpStatus.VARIATION_REQUESTED]
    current_task_type = Task.TaskType.AUTHORISE


class PreviewCoverLetterView(DocumentPreviewBase):
    document_types = [DocumentTypes.COVER_LETTER_PREVIEW, DocumentTypes.COVER_LETTER_PRE_SIGN]
    output_filename = "CoverLetter-Preview.pdf"


# This is new
class PresignCoverLetterView(PreviewCoverLetterView):
    current_status = [ImpExpStatus.PROCESSING, ImpExpStatus.VARIATION_REQUESTED]
    current_task_type = Task.TaskType.AUTHORISE


class PreviewCertificateView(DocumentPreviewBase):
    document_types = [DocumentTypes.CERTIFICATE_PREVIEW, DocumentTypes.CERTIFICATE_PRE_SIGN]
    output_filename = "Certificate-Preview.pdf"


# This is new
class PresignCertificateView(PreviewCertificateView):
    current_status = [ImpExpStatus.PROCESSING, ImpExpStatus.VARIATION_REQUESTED]
    current_task_type = Task.TaskType.AUTHORISE
```

It ensures the status / task type are correct.

Comment (MattHolmes123) [reply web/domains/case/views/views_pdf.py:67]: Then you'd need to change the three presign views in the urls.py file:
image

File: web/tests/conftest.py [hunk 4]
```diff
@@ -51,6 +51,7 @@
 from web.sites import SiteName
 from web.tests.helpers import CaseURLS, get_test_client
 
+from ..domains.case._import.fa_oil.models import ChecklistFirearmsOILApplication
```
Comment (MattHolmes123) [inline web/tests/conftest.py:54]: We refrain from using relative improt unless its the same directory.
`from web.domains.case` is preferred.

**Review by MattHolmes123 — COMMENTED** [review]
Added a few comments

File: web/tests/conftest.py (hunk 4 above)
Comment (MattHolmes123) [reply web/tests/conftest.py:54]: There are 6 x `from ..` and 1 x `from ...` in the codebase.
I have failed... :D

Comment (MattHolmes123) [reply web/tests/conftest.py:54]: Actually it should be this to match how we import all models.
`from web.models import ChecklistFirearmsOILApplication`

File: web/domains/case/views/views_pdf.py (hunk 3 above)
Comment (chris-pettinga) [reply web/domains/case/views/views_pdf.py]: Thanks for the detailed review, agreed I felt iffy about my implementation but didn't know about the `current_task_type` and `current_status` properties to restrict views, have implemented as such thank you.

Comment (chris-pettinga) [reply web/domains/case/views/views_pdf.py]: document_types are already checked here - https://github.com/uktrade/icms/blob/c69d8323df134036e484b12debaf01a4c25c8fc3/web/domains/case/views/views_pdf.py#L46-L47

File: web/domains/case/views/views_pdf.py [hunk 5]
```diff
@@ -68,16 +69,31 @@ class PreviewLicenceView(DocumentPreviewBase):
     output_filename = "Licence-Preview.pdf"
 
 
+class PresignLicenceView(PreviewLicenceView):
+    current_status = [ImpExpStatus.PROCESSING, ImpExpStatus.VARIATION_REQUESTED]
```
Comment (mikeylockwood) [inline web/domains/case/views/views_pdf.py:73]: Should the `document_types` for this just be `DocumentTypes.LICENCE_PRE_SIGN` with `DocumentTypes.LICENCE_PREVIEW` for `PreviewLicenceView`?

File: web/domains/case/views/views_pdf.py [hunk 6]
```diff
@@ -68,16 +69,31 @@ class PreviewLicenceView(DocumentPreviewBase):
...
+    current_task_type = Task.TaskType.AUTHORISE
+
+
 class PreviewCoverLetterView(DocumentPreviewBase):
     document_types = [DocumentTypes.COVER_LETTER_PREVIEW, DocumentTypes.COVER_LETTER_PRE_SIGN]
     output_filename = "CoverLetter-Preview.pdf"
 
 
+class PresignCoverLetterView(PreviewCoverLetterView):
+    current_status = [ImpExpStatus.PROCESSING, ImpExpStatus.VARIATION_REQUESTED]
```
Comment (mikeylockwood) [inline web/domains/case/views/views_pdf.py:83]: Same question as above regarding `document_types`

**Review by mikeylockwood — APPROVED** [review]

File: web/domains/case/views/views_pdf.py (hunk 5 above)
Comment (MattHolmes123) [reply web/domains/case/views/views_pdf.py:73]: I suggested the same but only as an afterthought in the initial comment e.g. "You could check the document_types as well."

**Review by MattHolmes123 — APPROVED** [review]

File: web/domains/case/views/views_pdf.py (hunk 3 above)
Comment (MattHolmes123) [reply web/domains/case/views/views_pdf.py]: Yer I know. I mean you could check for the specific one expected in that view (as Michael said in his comment).

Comment (chris-pettinga) [reply web/domains/case/views/views_pdf.py]: @MattHolmes123 sorry I think I get what you mean now, have re-pushed

File: web/domains/case/views/views_pdf.py (hunk 5 above)
Comment (chris-pettinga) [reply web/domains/case/views/views_pdf.py:78]: @mikeylockwood Yes I get what you mean now apologies, have re-pushed

File: web/domains/case/views/views_pdf.py [hunk 7]
```diff
@@ -64,20 +65,50 @@ def get(self, request: AuthenticatedHttpRequest, *args, **kwargs) -> HttpRespons
 
 
 class PreviewLicenceView(DocumentPreviewBase):
-    document_types = [DocumentTypes.LICENCE_PREVIEW, DocumentTypes.LICENCE_PRE_SIGN]
+    document_types = [
+        DocumentTypes.LICENCE_PREVIEW,
```
Comment (MattHolmes123) [inline web/domains/case/views/views_pdf.py:69]: Remove the `,` and black won't format this over three lines

**Pull request merged**